AndroidSDKCore/src/main/java/com/leanplum/Leanplum.java

```diff
@@ -1337,24 +1338,46 @@ public static void triggerMessageDisplayed(ActionContext actionContext) {
     ActionManager.getInstance().recordMessageImpression(actionContext.getMessageId());
     synchronized (messageDisplayedHandlers) {
       for (MessageDisplayedCallback callback : messageDisplayedHandlers) {
-        String messageID = actionContext.getMessageId();
-        String messageBody = "";
-        try {
-          messageBody = (String) actionContext.getArgs().get("Message");
-        } catch (Throwable t) {
-          Util.handleException(t);
-        }
-        String recipientUserID = Leanplum.getUserId();
-        Date deliveryDateTime = new Date();
-
-        MessageArchiveData messageArchiveData = new MessageArchiveData(messageID,
-                messageBody, recipientUserID, deliveryDateTime);
+        MessageArchiveData messageArchiveData = messageArchiveDataFromContext(actionContext);
         callback.setMessageArchiveData(messageArchiveData);
         OsHandler.getInstance().post(callback);
       }
     }
   }
 
+  private static MessageArchiveData messageArchiveDataFromContext(ActionContext actionContext) {
+    String messageID = actionContext.getMessageId();
+    String messageBody = "";
+    try {
+      messageBody = messageBodyFromContext(actionContext);
+    } catch (Throwable t) {
+      Util.handleException(t);
+    }
+    String recipientUserID = Leanplum.getUserId();
+    Date deliveryDateTime = new Date();
+
+    return new MessageArchiveData(messageID, messageBody, recipientUserID, deliveryDateTime);
+  }
+
+  @VisibleForTesting
+  public static String messageBodyFromContext(ActionContext actionContext) {
+    Object messageObject =  actionContext.getArgs().get("Message");
+    if (messageObject instanceof String) {
+      return (String) messageObject;
+    } else {
+      HashMap<String, String> messageDict = (HashMap<String, String>) messageObject;
+      if (messageDict.get("Text") != null &&
+              messageDict.get("Text") instanceof String) {
+        return messageDict.get("Text");
+      }
+      if (messageDict.get("Text value") != null &&
```

Comment:
I am curious how this Dictionary key was implemented with a space and caps "Text value" , we should definitely try to standardize this.

Reply:
this is the current key we're getting from the server. Agree on standardizing